src/connector/src/sink/log_store.rs

```diff
@@ -324,6 +333,252 @@ impl<R: LogReader> LogReader for MonitoredLogReader<R> {
+
+                // The first chunk after splitting will be returned
+                let first_chunk = chunks.next().unwrap();
+
+                // The remaining chunks will be pushed to the queue
+                for chunk in chunks.rev() {
+                    // The last chunk after splitting inherits the `is_last` from the original chunk
+                    self.core.unconsumed_chunk_queue.push_back(SplitChunk {
+                        chunk,
+                        upstream_chunk_offset,
+                        is_last,
+                    });
+                    is_last = false;
+                }
+
+                // Trigger rate limit and return the first chunk
+                let n = NonZeroU32::new(
+                    first_chunk.compute_rate_limit_chunk_permits(limit as _) as u32
+                )
+                .unwrap();
+                // chunks split should have effective chunk size <= limit
+                limiter.until_n_ready(n).await.unwrap();
+                SplitChunk {
+                    chunk: first_chunk,
+                    upstream_chunk_offset,
+                    is_last,
+                }
+            }
+        } else {
+            // Rate limit is not set
+            assert_eq!(self.rate_limit, None);
+            split_chunk
+        };
+
+        // Update the consumed_offset_queue if the `split_chunk` is the last chunk of the upstream chunk
+        let epoch = split_chunk.upstream_chunk_offset.epoch();
+        let downstream_chunk_id = self.core.next_chunk_id;
+        self.core.next_chunk_id += 1;
+        if split_chunk.is_last {
+            self.core.consumed_offset_queue.push_front((
+                TruncateOffset::Chunk {
+                    epoch,
+                    chunk_id: downstream_chunk_id,
+                },
+                split_chunk.upstream_chunk_offset,
+            ));
+        }
+
+        Ok((
+            epoch,
+            LogStoreReadItem::StreamChunk {
+                chunk: split_chunk.chunk,
+                chunk_id: downstream_chunk_id,
+            },
+        ))
+    }
+}
+
+impl<R: LogReader> LogReader for RateLimitedLogReader<R> {
+    async fn init(&mut self) -> LogStoreResult<()> {
+        self.core.inner.init().await
+    }
+
+    async fn next_item(&mut self) -> LogStoreResult<(u64, LogStoreReadItem)> {
+        let mut paused = false;
+        loop {
+            select! {
+                biased;
+                rate_limit_change = pin!(self.control_rx.recv()) => {
+                    if let Some(new_rate_limit) = rate_limit_change {
+                        let prev = self.update_rate_limit(new_rate_limit);
+                        paused = self.rate_limit == Some(0);
+                        tracing::info!("rate limit changed from {:?} to {:?}, paused = {paused}", prev, self.rate_limit);
+                    } else {
+                        tracing::warn!("rate limit control channel closed");
```

Comment:
Should the log reader exit here? Seems there is no way to restore the rate limit control channel.

Reply:
Simulation recovery test passed with `bail!` here.

Reply:
Is it okay that I commit the fix directly in this PR?